diff --git a/cpp/src/arrow/c/bridge.cc b/cpp/src/arrow/c/bridge.cc
index 6b30802c7..7c5dca04a 100644
--- a/cpp/src/arrow/c/bridge.cc
+++ b/cpp/src/arrow/c/bridge.cc
@@ -40,6 +40,7 @@
 #include "arrow/result.h"
 #include "arrow/stl_allocator.h"
 #include "arrow/type_traits.h"
+#include "arrow/util/align_util.h"
 #include "arrow/util/async_generator.h"
 #include "arrow/util/bit_util.h"
 #include "arrow/util/checked_cast.h"
@@ -2051,6 +2052,7 @@ Status ExportStreamNext(const std::shared_ptr<RecordBatchReader>& src, int64_t i
     ArrowArrayMarkReleased(out_array);
     return Status::OK();
   } else {
+    ARROW_ASSIGN_OR_RAISE(batch, util::EnsureAlignment(batch, 8, default_memory_pool()));
     return ExportRecordBatch(*batch, out_array);
   }
 }
